packages/next-server/server/render.tsx

```diff
@@ -20,6 +20,22 @@ type ComponentsEnhancer =
   | { enhanceApp?: Enhancer; enhanceComponent?: Enhancer }
   | Enhancer
 
+class ServerRouter extends Router {
+  constructor(pathname: any, query: any, asPath: any, ...args: []) {
+    super(pathname, query, asPath, ...args)
+    const coreMethodFields = ['push', 'replace', 'reload', 'back', 'prefetch', 'beforePopState']
+
+    coreMethodFields.forEach((field) => {
+      // @ts-ignore 'ServerRouter' has no index signature
+      this[field] = (...args: any) => {
+        const message = 'No router instance found.\n' +
+        'You should only use "next/router" inside the client side of your app.\n'
+        throw new Error(message)
+      }
+    })
+  }
+}
```

Comment:
I think something like this would be better as it doesn't define the methods on each constructor call:

```
function noRouter() {
  const message = 'No router instance found. you should only use "next/router" inside the client side of your app.'
  throw new Error(message)
}

class ServerRouter extends Router {
  push() {
    noRouter()
  }
  replace() {
    noRouter()
  }
  reload() {
    noRouter()
  }
  back() {
    noRouter()
  }
  prefetch() {
    noRouter()
  }
  beforePopState() {
    noRouter()
  }
}
```

I also think we should add a `err.sh` link for this error. I assume there's users running into this relatively often.

Reply:
Updated